Wallet Migration Wizard and integration

PartySensei opened the pull request:
- Implements `DlgWalletMigration.py` for migrating legacy wallets.
- Integrates migration flow with wallet creation wizard.
- Supports legacy, encrypted, and watching-only wallets.
- Provides user review, passphrase entry, and error handling.
- Stubs backend migration call (`migrate_wallet`) for future implementation.
- Follows Armory coding standards and UI guidelines.

Ready for review.

goatpig commented on ArmoryQt.py:
```diff
@@ -2749,31 +2749,51 @@ def execImportWallet(self):
    def execGetImportWltName(self):
-      fn = self.getFileLoad('Import Wallet File')
-      if not os.path.exists(fn):
+      """Get a wallet file to import and add it to the application."""
```
Undo all changes in ArmoryQt as a whole, the checks have to be done cpp side

goatpig commented on armoryengine/BDM.py:
```diff
@@ -11,6 +11,10 @@
 import random
 import threading
 import traceback
+import sys
```
What is this for?

goatpig commented on qtdialogs/qtdialogs.py:
```diff
@@ -22,6 +22,8 @@
 from zipfile import ZipFile, ZIP_DEFLATED
 
 from qtpy import QtCore, QtGui, QtWidgets
+from qtpy.QtCore import SIGNAL, SLOT  # Add PyQt4 signal/slot compatibility
```
Dont use SIGNAL/SLOT
Do not implement/fix GUI in qtdialogs.py, that file is only there for reference. Refactor the class out to its own file and progress from there.

goatpig commented on qtdialogs/qtdialogs.py:
```diff
@@ -390,17 +392,12 @@ def __init__(self, parent=None, main=None, initLabel=''):
       self.btnImportWlt = QtWidgets.QPushButton(self.tr("Import wallet..."))
-      self.connect(self.btnImportWlt, SIGNAL("clicked()"), \
-                    self.importButtonClicked)
+      self.connect(self.btnAdvCrypto, SIGNAL('toggled(bool)'), self.cryptoFrame, SLOT('setVisible(bool)'))
```
assign the slot directly to the signal instead:
`self.btnAdvCrypto.toggled.connect(setVisible)`

goatpig commented on qtdialogs/WalletMigrationSignals.py:
```diff
@@ -0,0 +1,68 @@
+#                                                                            #
+##############################################################################
+
+from qtpy import QtCore, QtWidgets
```
I dont understand the purpose of this file:
1. the signals are not used
2. there are unused imports
3. the passphrase reuse signal is silly, it can simply be checked when clicking next
4. the connectSignals method can implemented as is in DlgWalletMigration, why have an entire class for it, let alone a file?

PartySensei replied:
- New commit: Refactored DlgWalletMigration.py with improved UI consistency
- Removed WalletMigrationSignals.py as per maintainer feedback
- All previous feedback addressed

PartySensei committed:
Add wallet migration wizard UI

goatpig committed:
plug DlgWalletMigration to CppBridge methods

PartySensei committed:
Wallet Migration improvements to UI/UX, flow, error handling

goatpig commented on ArmoryQt.py:
```diff
@@ -1102,13 +1102,13 @@ def getExtraEntropyForKeyGen(self):
          pixDesk.save(pixBuf, 'PNG')
          source3 = bytes(pixBuf.buffer())
-      except:
+      except Exception as e:
```
don't capture the exception if you're not gonna use it

goatpig commented on qtdialogs/DlgUnlockWallet.py:
```diff
@@ -268,8 +271,13 @@ def changeScramble(self):
    def recycle(self):
-      QtWidgets.QMessageBox.critical(self, self.tr('Invalid Passphrase'), \
-         self.tr('That passphrase is not correct!'), QtWidgets.QMessageBox.Ok)
+      if getattr(self, '_suppressUnlockError', False):
```
Stop using getattr
On second thought, this much worse. Do not access child members from the parent!

goatpig commented on qtdialogs/DlgUnlockWallet.py:
```diff
@@ -287,8 +295,9 @@ def acceptPassphrase(self):
-      self.reply("")
-      self.reject()
+      # Only prompt the parent; do not reply or reject yet
+      if hasattr(self.parent, 'handleUnlockCancelForWatchOnly'):
```
and stop using hasattr

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -79,12 +82,14 @@ def __init__(self, parent, main):
-      if self.parent.reusePassprhase:
-         self._passphrase = passphrase
+      # Store the passphrase in the parent immediately after unlock
+      self.parent._unlockedPassphrase = passphrase
```
why store it in 2 different places?

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -168,193 +209,106 @@ def setupGridAndMinWidth(self, grid, labelTexts):
    def processUnlockRequest(self, ids):
-      #spawn unlock DlgUnlockWallet
+      # Ignore unlock requests if migration is already complete
+      if getattr(self, '_migration_complete', False):
```
what's going on here?

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -413,23 +367,86 @@ def processSetNewPassphrase(self, isPriv):
    def processCallback(self, payload):
       if payload.which() == 'cleanup':
+         self._migration_complete = True
+         if hasattr(self, 'dlgUnlock') and self.dlgUnlock:
```
stop using hasattr, forever.

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -413,23 +367,86 @@ def processSetNewPassphrase(self, isPriv):
+         else:
+            self._resultLabel.setStyleSheet("font-size: 12pt; font-weight: bold;" +
+               "color: #00FF00; background: transparent; border: none;")
+            self._resultLabel.setText("SUCCESS!")
```
f"Migration of Wallet {wltId} was successful!"

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -413,23 +367,86 @@ def processSetNewPassphrase(self, isPriv):
+         btnLayout.addStretch(1)
+         btnLayout.addWidget(self._doneBtn)
+         self.walletProgressPage.pageFrame.layout().addLayout(btnLayout)
+         self._doneBtn.show()
```
enable/disable the button rather than hide/show

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -413,23 +367,86 @@ def processSetNewPassphrase(self, isPriv):
+               self.stack.setCurrentWidget(self.setPassphrasePage)
+               self.resize(self.stack.currentWidget().sizeHint())
+            else:
+               self.reject()
```
reject with no notification to the user?
also no reply to the backend, it should be told there was a failure

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -532,117 +484,83 @@ def accept(self):
+         unlockDlg.reply("")
+         self.proceedWithMigration()
+      else:
+         self.reject()
```
same as the other reject, backend needs to know the migration process has to stop

PartySensei committed:
Adhered to maintainers feedback with few imporvements

goatpig commented on qtdialogs/DlgUnlockWallet.py:
```diff
@@ -343,3 +359,25 @@ def reply(self, passphrase):
       super().reply()
+
+################################################################################
+class DlgUnlockMigratingWallet(DlgUnlockWallet):
```
This class is a copy of the one in DlgMigrateWallet
It is never used, hence the `_suppressUnlockError` member being added into the parent class.

PartySensei committed:
Addressing maintainers feedback

PartySensei committed:
Reverted changes in qtdialogs/DlgUnlockWallet.py
Fixed child parent relationships, passphrase management, added back
DlgUnlockMigratingWallet class, clean up the code base in general

goatpig committed:
distinguish between empty passphrase and rejected unlock/set response

PartySensei committed:
Updated DlgMigration.py
Code updated post BE distinguishment introduction, added cancel buttons
for verify set new passphrase and verify passphrase, general code
cleanup

goatpig commented on qtdialogs/DlgWalletMigration.py:
```diff
@@ -618,3 +660,40 @@ def handleUnlockCancelForWatchOnly(self):
          # User does NOT want to create a watching-only wallet: close dialog, do not create wallet
          self.reject()
```
This reject does not notify migration end to the backend, but the one at line 520 does. Code at line 520 is repeated in `def closeEvent` overload.

You are better off overloading `reject` and implementing the cleanup there. So long as `migrationComplete` is not True and `reject` is called, you know you can reply to the backend with `success` set to False.

PartySensei replied:
Noted, fixing and will submit a new commit asap.

PartySensei committed:
Resolved code feedback, bug fixes and code cleanup
FE side now properly notifies migration end to BE
Fixed passhrase choice and reuse logic
Fixed UI refresh and dialog cancellation
Fixed code organization and some visual design elements

PartySensei committed:
Critical bug fixes and kdf params introduction
Added KDF parameter controls to migration wizard
Fixed processSetNewPassphrase logic issues
Enchanced passphrase workflow
Improved error handlding and user feedback
Code debug and clean up

goatpig committed:
detect and report network mismatch when importing legacy wallets
more thorough wallet file loading attempts
suppress redundant db loading warnings

goatpig committed:
fix highest used index in legacy wallet header
some import and style cleanup in DlgWalletMigration.py

PartySensei committed:
Refactor migration dialog: fix KDF config UI, WO wallet creation
Code cleanup, bug fixes, UI/UX improvements

PartySensei committed:
Enchance wallet migration dialog with improved field validation
Fixed invalid value display, added "last used" timestamp field and
dynamic button labeling

PartySensei committed:
Refactor of timestamp field

goatpig committed:
do not commit chaincode when migrating 1.35c roots
display human readable seed version in migration dialog
style and redundant code cleanup in DlgWalletMigration

PartySensei committed:
Fix wallet migration flow by removing blockign error dialogs

goatpig commented:
tACK
